### PR TITLE
set_link: Check the guest connective via ping host from guest

### PR DESCRIPTION
**Pull request by pingl2016**
Check the guest connective through ping host from guest instead of ping guest from host.

Signed-off-by: Ping Li <pingl@redhat.com>

### CODE REVIEW

**Comment by pingl2016** [comment]
ID: 1322010

File: qemu/tests/set_link.py [hunk 1]
```diff
@@ -77,48 +75,67 @@ def guest_reboot(reboot_method, link_up):
             if not link_up:
                 logging.info("Login error is expected when net link is down")
 
-    def guest_netwok_connecting_check(guest_ip, link_up, change_queues=False):
+    def ping_host_from_guest_check(session, link_up, guest_ifname=""):
```
Comment (xutian) [inline qemu/tests/set_link.py]: we have ping function in utils_net

File: qemu/tests/set_link.py [hunk 2]
```diff
@@ -77,48 +75,67 @@ def guest_reboot(reboot_method, link_up):
...
+            output = session.cmd_output(command, timeout=60)
+            if ttl in output:
+                break
+            time.sleep(2)
+        logging.info("ping statistics:\n '%s'" % output)
+        if not link_up and ttl in output:
             err_msg = "guest network still connecting after down the link"
             raise error.TestFail(err_msg)
-        elif link_up and utils_test.get_loss_ratio(output) == 100:
+        elif link_up and ttl not in output:
```
Comment (xutian) [inline qemu/tests/set_link.py]: why？

File: qemu/tests/set_link.py [hunk 3]
```diff
@@ -77,48 +75,67 @@ def guest_reboot(reboot_method, link_up):
...
-            bg_thread.join()
+    def guest_netwok_connecting_check(link_up, guest_ifname="",
+                                      change_queues=False):
+        """
+        Check whether guest network is connective by ping
+        """
+        session = vm.wait_for_serial_login()
+        try:
+            if change_queues:
+                queues = params.get("queues", 1)
```
Comment (xutian) [inline qemu/tests/set_link.py]: convert type here

File: qemu/tests/set_link.py (hunk 2 above)
Comment (pingl2016) [inline qemu/tests/set_link.py]: After set link off on windows, ping host will get "Destination net unreachable" and "0% loss". So use the ttl to check the connective of guest.

File: qemu/tests/set_link.py (hunk 1 above)
Comment (pingl2016) [inline qemu/tests/set_link.py]: Will Modify it.

**Comment by pezhang** [comment]
This case can not be successfully executed.  Hit below errors:
  File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/set_link.py", line 75, in guest_reboot
    TypeError: expected string or buffer

**Comment by pingl2016** [comment]
@pezhang From the log, the case failed after reboot vm. The patch has no effect on reboot. Anyway, I will debug it. Could you share the full log?

**Comment by pezhang** [comment]
@pingl2016  Below file contains the full log.
[debug.txt](https://github.com/autotest/tp-qemu/files/212230/debug.txt)

**Comment by pingl2016** [comment]
@pezhang There is a bug to track the issue you mentioned. Below is detail info.
https://bugzilla.redhat.com/show_bug.cgi?id=1323027
https://github.com/avocado-framework/avocado-vt/pull/430

**Comment by maxujun** [comment]
Acked-by:Xujun Ma xuma@redhat.com

**Comment by pezhang** [comment]
@pingl2016  

I retested with avocado-framework/avocado-vt#430 , still fail.
14:22:08 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/set_link.py", line 227, in run
14:22:08 ERROR|     change_queues)
14:22:08 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/set_link.py", line 182, in set_link_test
14:22:08 ERROR|     guest_reboot(reboot_method, link_up)
14:22:08 ERROR|   File "/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/set_link.py", line 75, in guest_reboot
14:22:08 ERROR|     vm.reboot(method=reboot_method, timeout=120)

Full log:

[debug.txt](https://github.com/autotest/tp-qemu/files/218522/debug.txt)

**Comment by pezhang** [comment]
This case hit below issue:
ID: 1296853

But this patch works, it can ping from guest to host.
Acked-by: Pei Zhang pezhang@redhat.com

**Comment by pingl2016** [comment]
@pezhang The issue you met is a known bug(ID: 1296853 ).  The patch just correct the guest connective check function which ping host from guest. From the log you attached, the patch has taken effect. If you found the case failed in other function, it may be a known bug or new bug. Thanks.

**Comment by suqinhuang** [comment]
Acked-by: Suqin Huang

File: qemu/tests/set_link.py [hunk 4]
```diff
@@ -195,6 +195,8 @@ def set_link_test(linkid, link_up, expect_status,
     change_queues = False
     guest_ifname = ""
     guest_ip = vm.get_address()
+    host_ip = utils_net.get_ip_address_by_interface(params.get("netdst"))
```
Comment (xutian) [inline qemu/tests/set_link.py]: why use srchost,  it looks srchost only used for migrate?
I think is enough.
host_ip  = utils_net.get_host_ip(params)

File: qemu/tests/set_link.py [hunk 5]
```diff
@@ -93,7 +91,8 @@ def guest_netwok_connecting_check(guest_ip, link_up, change_queues=False):
             utils_misc.wait_for(lambda: env["run_change_queues"], 30, 0, 2,
                                 "wait queues change start")
 
-        _, output = utils_test.ping(guest_ip, count=10, timeout=20)
+        _, output = utils_test.ping(host_ip, 10, interface=guest_ifname,
```
Comment (xutian) [inline qemu/tests/set_link.py]: How about ?
output = utils_test.ping(host_ip, count=10, timeout=20,
                                      interface=guest_ifname, session=session)[1]

Comment (pingl2016) [inline qemu/tests/set_link.py]: It's ok for me. Will update it.

File: qemu/tests/set_link.py (hunk 4 above)
Comment (pingl2016) [inline qemu/tests/set_link.py]: I have seen srchost defined in tests-example.cfg which represents the host ip address, and use function get_ip_address_by_interface() to make sure we can get host ip here. The second step become unnecessary actually. Will correct it as the comment you added.

**Comment by pingl2016** [comment]
@xutian patch has been updated.

**Pull request merged**